### PR TITLE
Bivariate legend

### PR DESCRIPTION
**Pull request by nerik**
Screenshot 2020-11-03 at 17 59 33

### CODE REVIEW

File: packages/ui-components/src/map-legend/Bivariate.tsx [hunk 1]
```diff
@@ -1,14 +1,177 @@
...
+  if (currentBucketIndex === undefined) {
+    currentBucketIndex = breaks.length
+  }
+  return currentBucketIndex
+}
+
+const getBivariateValue = (realValues: number[], breaks: number[][]) => {
+  if (realValues[0] === 0 && realValues[1] === 0) return 0
+  if (breaks) {
+    //  y: datasetB
```
Comment (j8seangel) [inline packages/ui-components/src/map-legend/Bivariate.tsx:48]: 🙇 thanks for the comments

File: packages/ui-components/src/map-legend/Bivariate.tsx [hunk 2]
```diff
@@ -1,14 +1,177 @@
...
+//    index returned is:            0 | 1 | 2 | 3 | 4 | 5
+//                                    |   |   |   |   |
+// Note: 0 is a special value, feature is entirely omitted
+//                                            |
+//                                       undefined
+const getBucketIndex = (breaks: number[], value: number) => {
+  let currentBucketIndex
+  for (let bucketIndex = 0; bucketIndex < breaks.length + 1; bucketIndex++) {
+    const stopValue =
+      breaks[bucketIndex] !== undefined ? breaks[bucketIndex] : Number.POSITIVE_INFINITY
```
Comment (j8seangel) [inline packages/ui-components/src/map-legend/Bivariate.tsx]: Did you know the [nullish coalescing operator](https://mariusschulz.com/blog/nullish-coalescing-the-operator-in-typescript) ? Do you think this is easier to read ?
```suggestion
    const stopValue = breaks?.[bucketIndex] ?? Number.POSITIVE_INFINITY
```

File: packages/layer-composer/src/types/index.ts [hunk 3]
```diff
@@ -37,13 +37,22 @@ export enum Group {
  */
 export interface LayerMetadataLegend extends GeneratorLegend {
   id?: string
-  type: 'colorramp' | 'bivariate' | 'solid'
+  type: 'colorramp' | 'colorramp-discrete' | 'bivariate' | 'solid'
   gridArea?: number
```
Comment (j8seangel) [inline packages/layer-composer/src/types/index.ts]: We have forgotten the `gridArea` property in heatmap-animated as it used to come from the stats, let's discuss if a key-value dictionary will be enough

File: packages/ui-components/src/map-legend/ColorRamp.tsx [hunk 4]
```diff
@@ -69,18 +82,32 @@ function ColorRampLegend({
             {currentValue}
           </span>
         )}
+        {type === 'colorramp-discrete' && (
+          <div className={styles.discreteSteps}>
+            {cleanRamp.map(([value, color], i) => (
+              <span
+                className={cx(styles.discreteStep, {
+                  [styles.discreteStepSelected]: currentBucketIndex === i,
```
Comment (j8seangel) [inline packages/ui-components/src/map-legend/ColorRamp.tsx]: Removed in the review with JC

File: packages/ui-components/src/map-legend/Bivariate.tsx [hunk 5]
```diff
@@ -1,14 +1,177 @@
...
 
 type BivariateLegendProps = {
-  layer: LegendLayer
+  layer: LegendLayerBivariate
   className?: string
 }
 
+/**
+ * ---------------------------------------------------------
+ * The following code is copy pasted from the MGL fork
```
Comment (j8seangel) [inline packages/ui-components/src/map-legend/Bivariate.tsx:14]: I can't think of a way of sharing this code easily, maybe duplicating and keep this comment is not that bad

**Review by j8seangel — COMMENTED** [review]
Amazing work! 👏

File: packages/layer-composer/src/generators/heatmap/util/get-legends.ts [hunk 6]
```diff
@@ -37,55 +40,93 @@ export const getColorRampBaseExpression = (config: GlobalHeatmapAnimatedGenerato
   return { colorRamp: colorRamps[0], colorRampBaseExpression: expressions[0] }
 }
 
+const STATS_MAX = 100
```
[the remark this answers is not in the document]
Comment (nerik) [reply packages/layer-composer/src/generators/heatmap/util/get-legends.ts]: I added some comments, we can discuss later

File: packages/layer-composer/src/types/index.ts [hunk 7]
```diff
@@ -37,13 +37,22 @@ export enum Group {
  */
 export interface LayerMetadataLegend extends GeneratorLegend {
   id?: string
-  type: 'colorramp' | 'bivariate' | 'solid'
+  type: 'colorramp' | 'colorramp-discrete' | 'bivariate' | 'solid'
```
[the remark this answers is not in the document]
Comment (nerik) [reply packages/layer-composer/src/types/index.ts:40]: Good idea 👍

**Pull request merged**